### PR TITLE
Add a playlist embed view

### LINKED ISSUE #360: Move /api/media/{key}/embed to UI
Embedding is fundamentally part of the UI and not the API. Move the existing embed view over to the UI. Still make use of the JWPlayer HTML5 player but configure it to use our API. Update the API to support MPEG-DASH and HLS. 

This is required by #359 since we need to move over to decoupling the player from the JWP delivery API.

Delivery API guide to sources: https://developer.jwplayer.com/jw-platform/docs/developer-guide/delivery-api/sources/

### PR DESCRIPTION
**Pull request by rjw57**
This PR is big. Sorry. I've spent some time trying to split it down into commits which can be understood individually. Generally there are individual commits which implement various bits of infrastructure and then a commit which makes use of it. If you review better seeing the use of components before looking at their implementation I recommend you review from the last commit back. If you review better by looking at components first and then seeing how they are put together, review from the first commit forward.

Note that GitHub is very poor at ordering commits properly. I recommend using local tooling.

This PR is so large because we change the way we handle embedding. Instead of using traditional Django templates, we render the embed view now like any other UI view. To this end the UI implements a specialised API which can be used to get configuration for the JWPlayer JavaScript API. This API is used by the frontend to configure the JWPlayer API manually rather than relying on JWPlatform to do it for us.

This gives us a nice degree of flexibility in how we structure the UI and lets us add our own playlist UI in the embed view.

Some time has been taken to perform JWPlayer integration "the React way" so that the various bits of the UI are decoupled from each other.

Since the media and playlist embed pages are so similar, a lot of the shared functionality has been broken out into generic components along with specialisations.

I've taken some care to write long commit messages for the trickier commits and tried to document what is going on through comments but ultimately some of this code is quite complex and may take a little time to get one's head around. I'm sorry for this but I don't think the problem can be solved in a simpler way without either a) copying and pasting a lot of code between the playlist and media embed pages or b) having to have everything JWPlayer related in a single "mega component".

If you are having trouble following what a component does, I suggest seeing how it is used in the rest of the code.

I've tested the playlist UI in Firefox on my machine and in IE11 on Windows 7 via browser stack.

It is also worth testing that the legacy embeds have not broken.

Closes #356. Closes #360.

### CODE REVIEW

File: ui/frontend/src/components/EmbedPlaylist.tsx [hunk 1]
```diff
@@ -0,0 +1,161 @@
+// Embedded playlist components.
```
[the remark this answers is not in the document]
Comment (rjw57) [reply ui/frontend/src/components/EmbedPlaylist.tsx]: It's at the top of the file.

File: ui/frontend/src/components/EmbedMessageDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
+import * as React from 'react';
```
[the remark this answers is not in the document]
Comment (rjw57) [reply ui/frontend/src/components/EmbedMessageDialog.tsx]: It's at the top of the file.

File: ui/serializers.py [hunk 3]
```diff
@@ -177,3 +178,40 @@ class PlaylistRSSSerializer(serializers.Serializer):
...
+        Ideas about whether a player served from http://localhost/ can access https URLs.
+
+        """
+        url = urllib.parse.urlparse(
+            jwplatform.pd_api_url(f'/v2/media/{obj.jwp.key}', format='json')
+        )
+        return urllib.parse.ParseResult('', *url[1:]).geturl()
+
+
+class JWPlayerConfigurationSerializer(serializers.Serializer):
```
[the remark this answers is not in the document]
Comment (rjw57) [reply ui/serializers.py]: I've had a bit of a ponder, and I think "configuration" is the "right" term. I'll add some clarifying comments though.

File: ui/urls.py [hunk 4]
```diff
@@ -64,4 +64,7 @@
 
     path('', TemplateView.as_view(template_name="index.html"), name='home'),
     path('search', TemplateView.as_view(template_name="index.html"), name='search'),
+
```
[the remark this answers is not in the document]
Comment (rjw57) [reply ui/urls.py]: Addressed in 5a738bc7a2629467d0670dfc5a22e942ac77a8af

File: ui/frontend/src/getJWPlayerStatic.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+ * Return a Promise which is resolved with the static JWPlayer API object. For example, to register
+ * a div with the id "myPlayer":
+ *
+ * ```js
+ * getJWPlayerStatic().then(jwplayer => jwplayer('myPlayer').setup({ ... }));
+ * ```
+ *
+ * The arguments to pass to ``setup()`` are documented in the JWPlayer API reference[1].
+ *
+ * Behind the scenes, this function will determine if the JWPlayer library hasl aready been loaded
```
[the remark this answers is not in the document]
Comment (rjw57) [reply ui/frontend/src/getJWPlayerStatic.ts]: Addressed in 4e430aab506799ef5c31765fa4ef1e176c4c31fc

File: ui/serializers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rjw57) [reply ui/serializers.py]: Comment updated in ffa8469792b0e5b33c13656750c7e02869e119f3

File: ui/frontend/src/components/EmbedMessageDialog.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (rjw57) [reply ui/frontend/src/components/EmbedMessageDialog.tsx]: Addressed by 4871409

**Comment by msb** [comment]
Tested on https://development.media.cam.ac.uk

**Pull request merged**